### DISCUSSION TITLE
P6Spy support

### DISCUSSION THREAD

**Discussion by donbeave** [open-telemetry/opentelemetry-java-instrumentation#3309] [Q&A]
Hello!

Do you plan to implement support for P6Spy? Probably can port p6spy for OpenTracing: https://github.com/opentracing-contrib/java-p6spy. If yes, I will be happy to do so.

Best regards,
  Alexey Zhokhov.

**Comment by trask** [comment]
This could be library instrumention for jdbc, cc: @anuraaga @mateuszrzeszutek

**Reply by donbeave** [reply]
Or probably something like `opentracing-jdbc`: https://github.com/opentracing-contrib/java-jdbc, which use the similar concept as `opentracing-p6spy` - use custom JDBC driver.

In my case the current implementation of `opentelemetry-java-instrumentation` for jdbc (https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jdbc) doesn't work as it uses bytecode transformation with Byte Buddy, which doesn't work with GraalVM.

**Reply by mateuszrzeszutek** [reply]
I really like the custom driver approach - anybody could use that, even applications that do not use p6spy.

**Reply by donbeave** [reply]
Make sense for me. I can port `opentracing-java-jdbc`.

**Reply by donbeave** [reply]
@anuraaga please let me know if you have any suggestions

**Reply by trask** [reply]
@donbeave I think you're good to go with the custom driver approach (`opentracing-java-jdbc`) 👍

**Reply by anuraaga** [reply]
I've personally only ever used instrumentation of the official drivers for example

https://github.com/openzipkin/brave/tree/master/instrumentation%2Fmysql8

If generic jdbc instrumentation can accomplish the same than that sounds good but I wonder if there are any gaps. I don't think we'd want something generic if it requires having these specific ones anyways to fill any gaps, but I've never compared the approaches.

**Reply by mateuszrzeszutek** [reply]
We're instrumenting a generic JDBC API in the javaagent too, so I'd expect the library instrumentation to have exactly the same functionality.

**Reply by anuraaga** [reply]
Cool - yeah then this sounds like a nice approach

**Reply by jasondlee** [reply]
@donbeave I'm interested in a port of opentracing-java-jdbc, as well. Let me know how I can help. If you need/want help. :)

**Reply by donbeave** [reply]
@jasondlee I'm working on it right now, basically, it's already ported to OTEL: https://github.com/donbeave/opentelemetry-jdbc. But didn't port tests yet. It's a very early preview, just the first draft version. But I want to replace most of the utility classes and reuse the code from the current javaagent implementation: https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jdbc/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jdbc

My plan is to make a library module and move some reusable classes from javaagent module to it.

I'm open to any help. You can add me on Telegram @donbeave and we can discuss how to split some work. Currently, I got workable jdbc implementation (which actually just a ported opentracing-jdbc library). And now I'm analyzing the structure of the current opentelemetry jdbc javaagent library to understand what can be reused and what can be shared between these two libraries.

**Comment by jasondlee** [comment]
@donbeave That's awesome. I thought it might take longer. :P 

I'm about to be out for a week, but I'll touch base when I get back to see where things stand and where I can pitch in. I will, though, clone your repo and give it a look in the time I have left today. :)
